crates/bevy_ecs/src/reflect/mod.rs

```diff
@@ -59,20 +61,34 @@ fn from_reflect_or_world<T: FromReflect>(
         return value;
     }
 
-    // Clone the `ReflectFromWorld` because it's cheap and "frees"
-    // the borrow of `world` so that it can be passed to `from_world`.
-    let Some(reflect_from_world) = registry.get_type_data::<ReflectFromWorld>(TypeId::of::<T>())
-    else {
+    fn different_type_error<T>(reflected: &str) -> ! {
         panic!(
-            "`FromReflect` failed and no `ReflectFromWorld` registration found for `{}`",
+            "the registration for the reflected `{}` trait for the type `{}` produced \
+            a value of a different type",
+            reflected,
             // FIXME: once we have unique reflect, use `TypePath`.
             std::any::type_name::<T>(),
         );
-    };
+    }
 
-    let Ok(mut value) = reflect_from_world.from_world(world).take::<T>() else {
+    let mut value = if let Some(reflect_default) =
+        registry.get_type_data::<ReflectDefault>(TypeId::of::<T>())
+    {
+        reflect_default
+            .default()
+            .take::<T>()
+            .unwrap_or_else(|_| different_type_error::<T>("Default"))
+    } else if let Some(reflect_from_world) =
+        registry.get_type_data::<ReflectFromWorld>(TypeId::of::<T>())
+    {
+        reflect_from_world
+            .from_world(world)
+            .take::<T>()
+            .unwrap_or_else(|_| different_type_error::<T>("FromWorld"))
+    } else {
         panic!(
-            "the `ReflectFromWorld` registration for `{}` produced a value of a different type",
+            "`FromReflect::from_reflect` failed and no registration for the reflected \
+            `Default` or `FromWorld` traits was found for the type `{}`",
```

Comment:
Self-review:
would it be better to not mention the registration at all, and just say that the type didn't reflect neither `Default` nor `FromWorld`? In an ideal world most people would just reflect traits and not know what a registration is, so telling them they need to reflect one of those traits would be easier to understand.

Reply:
I think it's fine to keep the mentions of "registration" as it could potentially be a soft introduction into registrations for users who are unaware of them.

Reply:
When using `ReflectFromReflect` this became a little harder to express because now there's another possible failure point (`ReflectFromReflect` being registered but failing when invoked), and I didnt' find a way to keep the reference to "registration".